plugins/SitesManager/SiteContentDetection/SiteContentDetectionAbstract.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Plugins\SitesManager\SiteContentDetection;
+
+use Piwik\SiteContentDetector;
+
+abstract class SiteContentDetectionAbstract
+{
+    public const TYPE_TRACKER = 1;
+    public const TYPE_CMS = 2;
+    public const TYPE_JS_FRAMEWORK = 3;
+    public const TYPE_CONSENT_MANAGER = 4;
+    public const TYPE_OTHER = 99;
+
+    public function __construct()
+    {
+    }
+
+    /**
+     * Returns the ID of the current detection. Automatically built from the class name (without namespace)
+     *
+     * @return string
+     */
+    public static function getId(): string
```

Comment:
Yeah, I'd agree here, the inheriting class could define the id. It can even be a `public static $id` property so that we can directly use it instead of calling a method (probably minor to no performance difference).

Reply:
I don't like using static properties for stuff that needs to be defined/overwritten in each subclass. This makes it a lot harder for a developer to understand it, without adding comments. Abstraction is meant to be used for this purpose. So either we leave it the way it is, or we make that method abstract and have it defined in each subclass.
And the purpose of having it automatically defined based on the class name ensures that the id is unique - at least for the stuff defined in core. Non unique IDs would cause a detection to be overwritten and to become unavailable.

Reply:
I used to use that a lot in the past as well. But at some point I tried starting to use abstract methods instead, so others that need to reuse the code have it easier and can't create problems by accident without even knowing it. If there would be something like abstract attributes, that would be what I would prefer most 😆